api/utils/keys/hardwaresigner.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2022 Gravitational, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keys
+
+import (
+	"bytes"
+	"crypto"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gravitational/trace"
+
+	attestation "github.com/gravitational/teleport/api/gen/proto/go/attestation/v1"
+)
+
+// HardwareSigner is a crypto.Signer which can be attested as being backed by a hardware key.
+// This enables the ability to enforced hardware key private key policies.
+type HardwareSigner interface {
+	crypto.Signer
+
+	// GetAttestationRequest returns an AttestationRequest for this private key.
+	GetAttestationRequest() (*AttestationRequest, error)
+
+	// GetPrivateKeyPolicy returns the PrivateKeyPolicy supported by this private key.
+	GetPrivateKeyPolicy() PrivateKeyPolicy
+}
+
+// GetAttestationRequest returns an AttestationRequest for the given private key.
+// If the given private key does not have a HardwareSigner, then a nil request
+// and error will be returned.
+func GetAttestationRequest(priv *PrivateKey) (*AttestationRequest, error) {
+	if attestedPriv, ok := priv.Signer.(HardwareSigner); ok {
+		return attestedPriv.GetAttestationRequest()
+	}
+	// Just return a nil attestation request and let this key fail any attestation checks.
+	return nil, nil
```

Comment:
Because a nil attestation statement is a valid response for non-hardware keys. We could change this to an error, but we'd always ignore the error, so I think this makes more sense with the current implementation. It's also noted in the Godoc so it shouldn't be a surprise to callers.